aten/src/ATen/native/cuda/RNN.cu

```diff
@@ -0,0 +1,601 @@
+#include "ATen/ATen.h"
+#include "ATen/AccumulateType.h"
+#include "ATen/TensorUtils.h"
+#include "ATen/NativeFunctions.h"
+#include "ATen/cuda/CUDAContext.h"
+#include "ATen/cuda/CUDAApplyUtils.cuh"
+#include <THC/THCNumerics.cuh>
+
+namespace at { namespace native {
+
+namespace {
+
+using at::cuda::detail::TensorInfo;
+using at::cuda::detail::getTensorInfo;
+using at::cuda::detail::IndexToOffset;
+using at::cuda::detail::canUse32BitIndexMath;
+
+// Factor will be 3 for GRU and 4 for LSTM
+void checkSizes(CheckedFrom c,
+                const TensorArg& input_gates, const TensorArg& hidden_gates,
+                const TensorArg& input_bias, const TensorArg& hidden_bias,
+                int64_t factor, const TensorArg& prev_hidden) {
+  checkDim(c, input_gates, 2);
+  checkSameSize(c, input_gates, hidden_gates);
+  int64_t gates_size = input_gates->size(1);
+
+  if (input_bias->defined()) {
+    checkDim(c, input_bias, 1);
+    checkNumel(c, input_bias, gates_size);
+    checkSameSize(c, input_bias, hidden_bias);
+  }
+
+  checkDim(c, prev_hidden, 2);
+  checkNumel(c, prev_hidden, input_gates->size(0) * gates_size / factor);
+
+  checkAllSameGPU(c, {input_gates, hidden_gates, input_bias, hidden_bias, prev_hidden});
+}
+
+bool allContiguous(at::TensorList tensors) {
+  return std::all_of(tensors.begin(), tensors.end(),
+                     [](const at::Tensor& t) { return !t.defined() || t.is_contiguous(); });
+}
+
+void getLaunchConfig(dim3& block, dim3& grid, int64_t numel) {
+  int curDevice = -1;
+  cudaGetDevice(&curDevice);
+  block = cuda::getApplyBlock();
+  AT_ASSERTM(cuda::getApplyGrid(numel, grid, curDevice),
+             "Could not get grid size for pointwise apply.");
+}
+
+template<typename T, typename T2>
+TensorInfo<T, T2> tryGetTensorInfo(const at::Tensor& t) {
+  return t.defined() ? getTensorInfo<T, T2>(t) : TensorInfo<T, T2>{};
+}
+
+void collapseDims() {};
+template<typename T, typename T2, typename... Args>
+void collapseDims(TensorInfo<T, T2>& info, Args&... infos) {
+  info.collapseDims();
+  collapseDims(infos...);
+}
+
+#define DEVICE_LINEAR_GET(D_TENSOR, INDEX)                              \
```

Comment:
Do these still have to be macros? Can they be functions?

Reply:
I would like to have them as functions, but they would still need those template arguments (they can't really be inferred from the argument types), and so I left them as macros.